### PR TITLE
Feature: shut down the computer at or after a certain time

### PR DESCRIPTION
**Pull request by luguina**
A new `-shutdown [TIME]` option has been created to specify the time when the client close and will shut down the host computer. This is useful functionality for clients running in cloud environment where the compute capacity is metered. The time argument can have two different formats: an absolute date & time in the format _yyyy-MM-ddTHH:mm:ss_ (24h format) or relative time in the format _+m_ where m is the number of minutes from now. The minimum scheduled time is 10 minutes from the moment the client is invoked. 

The user can also select the `-shutdown-mode [MODE]`, where _MODE_ can take two possible values: wait or hard. "Wait" will wait for all the processes (render + pending uploads) to finish and "hard" will cancel all the pending jobs, close the client and will initiate the computer shutdown. The default shutdown mode is _wait_.

### CODE REVIEW

File: src/com/sheepit/client/os/OS.java [hunk 1]
```diff
@@ -45,6 +45,8 @@ public String getCUDALib() {
 	
 	public abstract boolean checkNiceAvailability();
 	
+	public abstract void shutdownComputer();
```
Comment (intrigus) [inline src/com/sheepit/client/os/OS.java:48]: Why does this not get passed the delay value?
Imho we should pass the delay here and not specify the delay in every os specific function.

File: src/com/sheepit/client/os/Windows.java [hunk 2]
```diff
@@ -227,4 +227,16 @@ int getPriorityClass(int priority) {
...
+	
+	@Override public void shutdownComputer(int delayInMinutes) {
+		try {
+			// Shutdown the computer, waiting 60 seconds, force app closure and on the shutdown screen indicate that was initiated by SheepIt app
+			ProcessBuilder builder = new ProcessBuilder("shutdown", "/s", "/f", "/t", String.valueOf(delayInMinutes * 60), "/c", "\"SheepIt App has initiated this computer shutdown.\"");
+			Process process = builder.inheritIO().start();
+		}
+		catch (IOException e) {
+			System.err.println(
+					String.format("Windows::startShutdown Unable to execute the command 'shutdown /s /f /t 60...' command. Exception %s", e.getMessage()));
```
Comment (laurent-clouet) [inline src/com/sheepit/client/os/Windows.java:239]: Shouldn't it be String.format("Windows::shutdownComputer  ?

Comment (luguina) [reply src/com/sheepit/client/os/Windows.java:239]: It does. IntelliJ refactoring option is not as "Intelli" as that :). I'll fix it.

File: src/com/sheepit/client/Client.java [hunk 3]
```diff
@@ -237,7 +276,11 @@ public void run() {
 						this.gui.status(String.format("Cannot connect to the server. Please check your connectivity. Will try again at %tR",
 								new Date(new Date().getTime() + time_sleep)));
 						try {
-							Thread.sleep(time_sleep);
+							int timeSlept = 0;
+							while (timeSlept < time_sleep && this.running && !this.shuttingdown) {
```
Comment (intrigus) [inline src/com/sheepit/client/Client.java:280]: Maybe extract this to a method?

**Pull request merged**